integration_tests/run.sh

```diff
@@ -28,61 +28,221 @@
+     "git_author=`git show -s --format="%ae" "$GIT_COMMIT"`" "p2g_package_published=FALSE"\
+     "g2g_package_published=FALSE" "GIT_TAG_NAME=$GIT_TAG_NAME" "commit_message=$commit_message"
+
+
+    # If the tag is correct and a version was specified, make a version release.
+    if [[ "$GIT_TAG_NAME" =~ ^[pg]2g-v[0-9]+\.[0-9]+\.[0-9]+ ]]; then
+        # Removes prefix from $GIT_TAG_NAME.
+        suffix="${GIT_TAG_NAME:5}"
+    fi
+
+    if [[ "${GIT_TAG_NAME:0:3}" = "g2g" ]] || [[ "$commit_message" =~ (^|.[[:space:]])"["g2g(-skip-tests)?"]"$ ]]; then
+        prefixes=geo
+        save_vars "p2g_tests=SKIPPED" "p2g_documentation=SKIPPED" "p2g_package="\
+         "g2g_tests=FAILED" "g2g_documentation=FAILED" "g2g_package=geo2grid-${suffix}"
+    elif [[ "${GIT_TAG_NAME:0:3}" = "p2g" ]] || [[ "$commit_message" =~ (^|.[[:space:]])"["p2g(-skip-tests)?"]"$ ]]; then
+        prefixes=polar
+        save_vars "p2g_tests=FAILED" "p2g_documentation=FAILED" "p2g_package=polar2grid-${suffix}"\
+         "g2g_tests=SKIPPED" "g2g_documentation=SKIPPED" "g2g_package="
+    else
+        prefixes="geo polar"
+        save_vars "p2g_tests=FAILED" "p2g_documentation=FAILED" "p2g_package=polar2grid-${suffix}"\
+         "g2g_tests=FAILED" "g2g_documentation=FAILED" "g2g_package=geo2grid-${suffix}"
+    fi
+}
+
+setup_conda()
+{
+    # Activate conda for bash.
+    /data/users/davidh/miniconda3/bin/conda init bash
+    # Restart the shell to enable conda.
+    source ~/.bashrc
+
+    conda env update -n jenkins_p2g_swbundle -f "$WORKSPACE"/build_environment.yml
+    # Documentation environment also has behave, while the build environment does not.
+    conda env update -n jenkins_p2g_docs -f "$WORKSPACE"/build_environment.yml -f "${WORKSPACE}/jenkins_environment.yml"
     conda activate jenkins_p2g_docs
-    if [[ ! "$commit_message" =~ (^|[[:space:]])"["([pg]2g-)?skip-tests"]"$ ]]; then
+    pip install -U --no-deps "$WORKSPACE"
+}
+
+format_test_details()
+{
+    test_details="$1"
+    json_file="${WORKSPACE}/integration_tests/json_file.txt"
+    # Gets the line before json data starts.
+    i=`grep -n "^\[$" "$test_details" | grep -oE "[0-9]+"`
+    # Gets the line after json data ends.
+    j=`grep -n "^\]$" "$test_details" | grep -oE "[0-9]+"`
+    # Remove lines that are not json data.
+    sed "1,${i}d;${j},\$d" "$test_details" > "$json_file"
+    set +x
+    python << EOF > "$test_details"
+import json
+with open("json_file.txt") as json_file:
+    data = json.load(json_file)
+    print()
+    for test in data['elements']:
+        name = test['name'].split('@')[1]
+        duration = 0
+        for step in test['steps']:
+            duration += step['result']['duration'] if step.get('result') else 0
+        end = '\n'
+        if test == data['elements'][-1]:
+            end = ''
+        print("\t\t{0}: {1} in {2} seconds".format(name, test['status'], round(duration)), end=end)
+EOF
+    set -x
+    rm "$json_file"
+}
+
+run_tests()
+{
+    try
```

Comment:
What is the `try` function if you just reset the `set -e` call in the subshell? What is the `try` function doing that isn't equivalent to calling `set +e`?

Reply:
I agree that using too many subshells is confusing. I tried to limit their use and instead use OR and AND statements where needed to simplify the logic.